include/swift/AST/DiagnosticConsumer.h

```diff
@@ -95,10 +95,10 @@ class DiagnosticConsumer {
   ///
   /// \param Info Extra information associated with the diagnostic.
   virtual void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
-                                DiagnosticKind Kind,
-                                StringRef FormatString,
+                                DiagnosticKind Kind, StringRef FormatString,
                                 ArrayRef<DiagnosticArgument> FormatArgs,
-                                const DiagnosticInfo &Info) = 0;
+                                const DiagnosticInfo &Info,
+                                StringRef currentPrimaryInput) = 0;
```

Comment:
Hm. I feel really weird having a *DiagnosticConsumer* know about "primary inputs". DiagnosticConsumer is a general interface, and not everything we do has a particular primary file associated with it. Moreover, "primary files" are a Frontend concept, and DiagnosticConsumer is down in AST, so this smells like a conceptual layering violation to me. (Also, I can't remember if whole-module mode uses "primary files" at all.)

An alternate idea I had was to use a SourceLoc as the "current" token, which for now could be the start of the "current" file. I still don't really know what to call it, though.

Reply:
Thank you for your clear summary of our talk.

I like your thinking about the layering. If we pass in a `SourceLoc`, we could call it the `diagnosticDestinationOfLastResort`. Or the `finalRestingPlace`. 

OTOH, it might seem unintuitive to use a `SourceLoc` as a stand-in for the `.dia` file. Here's another idea: we could pass in the `.dia` file of the current primary, but then FileSpecificDiagnosticConsumer would need another (unordered) map. What do you think? Any other ideas?

Reply:
Another thought, we could call the thing that's passed in to the `DiagnosticEngine`, by it filename or location, the `diagnosticOrphanage`, since it's where homeless ones go. `diagnosticMotel`? (Diagnostics check in but the don't check out.) "Refuge"? I hope all these suggestions will inspire you to think of something really good!

Reply:
Let me try it with a location and see how it looks

Reply:
I really want to keep knowledge of FileSpecificDiagnosticConsumer out of DiagnosticEngine as much as possible, mostly to preserve separation of concerns. It already feels weird to me to have it track which file is currently being processed.

`fallbackBufferLocation`? (I put "buffer" in there to make it clear that the exact SourceLoc isn't relevant.) `activeBuffer`? `bufferIndirectlyResponsibleForDiagnostic`?

Reply:
These are good! I'll ponder. Thanks!

Reply:
Another design alternative would be to not store the fallbackBufferLocation in the DiagnosticEngine, but for it to merely forward the set method to all consumers. The FileSpecificDiagnosticConsumer would be the only one to store it. How does that alternative feel to you?